One app to rule them all

whikloj opened the pull request:
**GitHub Issue**: https://github.com/Islandora/documentation/issues/731

# What does this Pull Request do?

Moves Alpaca from an OSGI bundle to a single runnable jar with all the various services inside.

# What's new?

* Removes OSGI bindings
* Pulls SparqlUpdateProcessors in from fcrepo-camel code (it was causing problems with Jena libraries otherwise)
* Create single runnable jar that takes configuration from a properties file.

* Does this change require documentation to be updated? Docs are updated
* Does this change add any new dependencies? No
* Does this change require any other modifications to be made to the repository (i.e. Regeneration activity, etc.)? No
* Could this change impact execution of existing code? Deployment will need to be altered.

# How should this be tested?

Using a playbook or ISLE instance, build the jar and then copy it into the system (somewhere it can access Fedora, Blazegraph, ActiveMQ, the microservices). 

Shutdown Karaf and copy the settings from all the various configuration files to a single properties file. 

Start the jar with that properties file and it should perform all the various actions.

ie.
```shell
java -jar islandora-alpaca-app-2.0.0-all.jar -c config.properties
```

# Additional Notes:

There are some stacktraces thrown that I left in place but are not actually a problem (for instance when you choose a media file, Drupal triggers us to send an event but it doesn't have all the necessary URLs yet. It can be ignored but could also be a sign of a problem.

Could leave the less verbose in the INFO level and move the stacktrace to the DEBUG logging level, but I don't want anyone to miss an actual problem. 🤷 

# Interested parties
@Islandora/8-x-committers (especially those you don't hate Java or at least will do Java, @dannylamb, @elizoller, @adam-vessey, etc)

whikloj committed:
Make a single jar with configurable services

whikloj committed:
Add example properties

whikloj committed:
Remove old property files, fix license on classes copied from Fedora

whikloj committed:
More fixes to README

whikloj committed:
Fixes from tests not working

whikloj committed:
Logging for custom http client settings

whikloj committed:
Alter README

whikloj committed:
Alter Github Actions to use Java 11

elizoller commented:
can you be more specific about 
>Start the jar with that properties file
?
like literally `java -jar ...` or do i need to run it some other way?

whikloj commented:
Sorry @elizoller I didn't realize I hadn't put that in here.

I did update the README which (hopefully) has all the needed information. https://github.com/Islandora/Alpaca/blob/d92362fd418cd5ad2d306ef7901a536bd555907d/README.md

adam-vessey commented on README.md:
````diff
@@ -10,11 +10,190 @@ Event-driven middleware based on [Apache Camel](http://camel.apache.org/) that s
+```
+This defines the pool of connections to the ActiveMQ instance.
+```
+jms.concurrent-consumers=1
````
Not sure how exactly this is applied... in the Karaf world, I believe this was applied per-connector... in this new jar form, is this (one consumer) the same across all queues? I can see utility in being able to adjust per-queue, as some microservices (Hypercube/Tesseract OCR stuff) are themselves multithreaded/-processed... so an OCR consumer would create more load than a Houdini/imagemagick consumer... being able to balance/limit things?

... on the other hand, could presumably just spin up separate JAR instances for each service, if one wanted them to be configured differently...

whikloj replied:
It is a shared connection pool for all the various services and so they all share the concurrent consumer configuration. 

I could remove this and instead add a property for each service (ie. `fcrepo.indexer.concurrent-consumers`, `derivatives.ocr.concurrent-consumers`, etc) which if set would then be set directly in the route.

Would that be preferable?

adam-vessey replied:
I believe so, thinking that such granularity would allow an additional level of tuning if so desired.

whikloj committed:
Make concurrent consumers configurable per service

whikloj committed:
Add disableStreamCache=true option to http calls

whikloj commented:
Sorry folks I decided to pull in the changes from #66 as well. This should be ready for a full test.

elizoller commented:
I feel like I can't really speak to the code part of this directly but I can provide feedback on what I did to test this PR.
I pulled in this PR in alpaca (in a playbook) and used the gradle command in the readme to build it. Then I started the jar with the commands provided in the readme with the provided examples.properties file. Then I went into the playbook and created some items and media. I verified the following functionality works: indexing items and media in fedora, deleting items removes them from fedora, updating an item creates a version in fedora, homarus/ffmpeg derivs generated, houdini/imagemagick derivatives generated, hypercube/tesseract/pdf full text extracted, crayfits/fits derivatives generated.

whikloj commented:
No one can speak to Java code. 

Like all code it is best glimpsed in one's peripheral vision. It is the fairy lights in the trees, the oasis on the horizon, it is real but subjective.

![and like that...he's gone](https://c.tenor.com/n56WNgm5WzkAAAAM/and-like-that-hes-gone-the-usual-suspects.gif)

seth-shaw-unlv commented on islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/FcrepoIndexer.java:
```diff
@@ -178,22 +168,32 @@ public void configure() {
+                .setBody(constant(null))
 
-                // Remove the file from Drupal.
-                .toD(getMillinerBaseUrl() + "node/${exchangeProperty.uuid}?connectionClose=true");
```
Question: why are we dropping the `?connectionClose=true`?

adam-vessey replied:
Haven't fully traced through; however, suspecting it might've just been moved... as in, the [`config.addHttpOptions()` thing called just below](https://github.com/Islandora/Alpaca/pull/82/files#diff-a51d1a13525c54cfeab1b9c8677bb4275ad4a90f01d809bd23b0a057e9c01540R176), which then adds it... assuming it's the method being called: https://github.com/Islandora/Alpaca/pull/82/files#diff-d16604429d3efb9fc5df0c659733a59a957a29fa11ce2309f44588b72865c547R105

seth-shaw-unlv replied:
There it is. I figured it was something like that.

adam-vessey replied:
... this said... the `.setHeader(Exchange.HTTP_URI, [...])` bit almost looks like it should also have `config.addHttpOptions()` wrapped around it? Or possibly additionally added in a `.setHeader(Exchange.HTTP_QUERY, [...])` kind of thing?

adam-vessey replied:
Actually... could this be premature optimization (moving the actual URI to the header, and specifying whatever "fake" endpoint in the `.toD()` business)? Looks like `camel-http` may (be able to) automatically handle things?: https://camel.apache.org/components/3.12.x/eips/toD-eip.html#_using_optimised_components

whikloj replied:
My understanding is that `connectionClose=true&disableStreamCache=true` are options for the http endpoint. 

I moved the full URL out of the `toD()` as using dynamic properties (ie. `${exchangeProperty.uuid}`) that change alot can be [resource intensive](https://camel.apache.org/components/3.12.x/eips/toD-eip.html#_avoid_creating_endless_dynamic_endpoints_which_takes_up_resources)

So I send the actual URL in the `Exchange.HTTP_URI` header and kept those options on the actual endpoint as they are unimportant to the actual target URL. 

However I can try tracing the http endpoint to see what happens because it is possible I have done it wrong.

seth-shaw-unlv replied:
Adding `?connectionClose=true` wasn't premature. We added it specifically because it wasn't closing the connections in a timely manner and I was hitting an absurd number of open connections when doing large migrations.

whikloj replied:
Ok I see what @adam-vessey is saying, my switch of moving the dynamic URI from the `toD` into the `Exchange.HTTP_URI` should be unnecessary as the http component should optimize the route to do that automatically.

I can switch it back

adam-vessey replied:
Yeah, that's what I'm thinking.

... Somewhat associated, but not sure how it handles parsing out to `Exchange.HTTP_QUERY` from the `.toD()` value, and recombining into the `Exchange.HTTP_URI` value, or if we would have to follow their pattern of just providing the `Exchange.HTTP_PATH` and `Exchange.HTTP_QUERY`... or providing the fully-built out URI in `Exchange.HTTP_URI`?... consideration of which should be somewhat side-stepped by allowing the component to do its own optimization.

seth-shaw-unlv commented:
Other than my question about connectionClose (see above), I don't see any obvious issues. I think we can probably trust your smoke tests on this, @elizoller.

seth-shaw-unlv commented:
Okay, as far as I'm concerned, if @elizoller is happy with his tests and @adam-vessey has his questions adequately addressed, 👍.

whikloj committed:
Switch back to using toD with full URL

whikloj committed:
Missed clean-up

whikloj commented:
Sorry for the delay on this, I have reverted the URL changes and then I had to work out a way of testing as the old way did not seem to work.

adam-vessey commented on islandora-connector-derivative/src/main/java/ca/islandora/alpaca/connector/derivative/DerivativeOptions.java:
```diff
@@ -0,0 +1,190 @@
+          // Add connectionClose and other http options.
+          final String finalOutput = addHttpOptions(output);
+          camelContext.addRoutes(new DerivativeConnector(serviceName, finalInput, finalOutput, this));
+        } catch (final Exception e) {
```
PMD's complaining about this:

```
1 	/home/adam/repos/Alpaca/islandora-connector-derivative/src/main/java/ca/islandora/alpaca/connector/derivative/DerivativeOptions.java 	106 	Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block
```

Linking to: https://pmd.github.io/pmd-6.21.0/pmd_rules_java_design.html#avoidcatchinggenericexception

whikloj replied:
Yeah, some of these PMD rules are suspect to me and might require some review. 

In this case we have no option as that is the error returned by `camelContext.addRoutes()`

https://github.com/apache/camel/blob/camel-3.5.0/core/camel-api/src/main/java/org/apache/camel/CamelContext.java#L570

adam-vessey replied:
Do... we want to catch it, or propagate it back up the chain? As in, being configured to start with the derivative connector, but unable to set things up... should possibly propagate back up to kill the app, to be more noticeable instead of just dumping the stacktrace but proceeding after the failure?

... That said, not terribly sure what the call stack would look like, when hitting this... from `setApplicationContext()`, and then into the depths of Spring?

whikloj replied:
I can let it flow back up to `setApplicationContext()` but then it looks like I need to wrap it in a `BeansException` to get it out.
Which I can do but then I'm still catching the `Exception` there and will need to ignore there.

adam-vessey replied:
Is this only happening because we're not quite matching up with the expectations of the Spring Bean lifecycle? As in: Our `setApplicationContext()` thing here isn't really just setting the context, it's more doing initialization things.

Really, could we not just have the `CamelContext` injected (possibly even in the constructor?), and then do our initialization business with it? ... thinking I might take a shot at this approach (and throw a PR against this one).

whikloj replied:
Go for it, I think I tried that but I could be wrong. I can't quite remember.

adam-vessey replied:
I've got https://github.com/whikloj/Alpaca/compare/one-app...adam-vessey:fixies

... which seems to work fine for the derivative side of things (that I've tested... really just service image and thumbnail stuff, but not anticipating it blowing up otherwise); however... in trying to test the Fedora stuff, having the exchange go sideways:

```
ERROR 17:53:00.693 [Camel (camel-1) thread #6 - seda://nodeIndex] (DefaultErrorHandler) Failed delivery for (MessageId: queue_islandora-indexing-fcrepo-content_ID_islandora8-46433-1638289432407-3_19_-1_1_1 on ExchangeId: ID-islandora8-1638381174585-0-5). Exhausted after delivery attempt: 6 caught: org.apache.camel.InvalidPayloadException: No body available of type: java.io.InputStream but has value: null of type: null on: JmsMessage[JmsMessageID: ID:islandora8-46433-1638289432407-3:19:-1:1:1]. Caused by: No type converter available to convert from type: null to the required type: java.io.InputStream with value null. Exchange[ID-islandora8-1638381174585-0-5]. Caused by: [org.apache.camel.NoTypeConversionAvailableException - No type converter available to convert from type: null to the required type: java.io.InputStream with value null]. Processed by failure processor: FatalFallbackErrorHandler[Channel[log12]]

Message History (complete message history is disabled)
---------------------------------------------------------------------------------------------------------------------------------------
RouteId              ProcessorId          Processor                                                                        Elapsed (ms)
[FcrepoIndexerNodeI] [FcrepoIndexerNodeI] [from[seda://nodeIndex]                                                        ] [      5270]

Stacktrace
---------------------------------------------------------------------------------------------------------------------------------------

org.apache.camel.InvalidPayloadException: No body available of type: java.io.InputStream but has value: null of type: null on: JmsMessage[JmsMessageID: ID:islandora8-46433-1638289432407-3:19:-1:1:1]. Caused by: No type converter available to convert from type: null to the required type: java.io.InputStream with value null. Exchange[ID-islandora8-1638381174585-0-5]. Caused by: [org.apache.camel.NoTypeConversionAvailableException - No type converter available to convert from type: null to the required type: java.io.InputStream with value null]
	at org.apache.camel.support.MessageSupport.getMandatoryBody(MessageSupport.java:117)
	at org.apache.camel.component.http.HttpProducer.createRequestEntity(HttpProducer.java:576)
	at org.apache.camel.component.http.HttpProducer.createMethod(HttpProducer.java:477)
	at org.apache.camel.component.http.HttpProducer.process(HttpProducer.java:118)
	at org.apache.camel.support.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:66)
	at org.apache.camel.processor.SendDynamicProcessor.lambda$process$0(SendDynamicProcessor.java:195)
	at org.apache.camel.impl.engine.DefaultProducerCache.doInAsyncProducer(DefaultProducerCache.java:317)
	at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:180)
	at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$RedeliveryTask.redeliver(RedeliveryErrorHandler.java:788)
	at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:148)
	at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:287)
	at org.apache.camel.component.seda.SedaConsumer.sendToConsumers(SedaConsumer.java:274)
	at org.apache.camel.component.seda.SedaConsumer.doRun(SedaConsumer.java:187)
	at org.apache.camel.component.seda.SedaConsumer.run(SedaConsumer.java:130)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
        [... and so on ...]
```

Seeing the change from `simple("${null}")` to `constant(null)`, poking around there, but really... unsure. Tests seem to go fine inside of Gradle itself, while building.... not sure if there might be some configuration that I'm missing somewhere or what...

adam-vessey replied:
Yeah... I think it's something related to the fact that it's being expressed as whatever expression?

```java
        msg.setBody(null);
```
in the `CommonProcessor` seemed to do the trick?; however... then seeing other stuff go sideways.

... as in, looking at fcrepo, seeing things go in like:

```
http://localhost:8080/fcrepo/rest/exchangeProperty(fedoraBaseUrl)/cd/a3/48/f0/cda348f0-5ee5-435f-b0e7-ff3249e10f72
```

That `exchangeProperty(fedoraBaseUrl)` thing... seems like something isn't being templated/replaced correctly? Also in that `CommonProcessor` thing?

whikloj replied:
Oh yeah, weird that it would work in tests but fail in real life. But it looks like I needed to reference the exchange as we came out.
https://github.com/whikloj/Alpaca/commit/2954e4d7d55d5de7489a8d99967f7f9d185e0bc2

whikloj replied:
I added a commit to fix those two things in CommonProcessor, so if you want to try that and then open your PR.

adam-vessey replied:
So yeah... suspecting that expressions are not expected to be used as they are in the `CommonProcessor` thing... setting explicit values, such as:

```diff
diff --git a/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java b/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
index 80c7b27..33b6063 100644
--- a/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
+++ b/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
@@ -17,9 +17,6 @@
  */
 package ca.islandora.alpaca.indexing.fcrepo;

-import static org.apache.camel.language.constant.ConstantLanguage.constant;
-import static org.apache.camel.language.property.ExchangePropertyLanguage.exchangeProperty;
-
 import org.apache.camel.Exchange;
 import org.apache.camel.Processor;

@@ -49,8 +46,8 @@ public class CommonProcessor implements Processor {
         final AS2Event json = msg.getBody(AS2Event.class);
         exchange.setProperty("fedoraBaseUrl", json.getTarget());
         msg.removeHeaders("*", "Authorization");
-        msg.setHeader(config.getFedoraUriHeader(), exchangeProperty("fedoraBaseUrl"));
-        msg.setBody(constant(""));
+        msg.setHeader(config.getFedoraUriHeader(), json.getTarget());
+        msg.setBody(null);
         exchange.setMessage(msg);
     }
 }
```

seems to do the trick. The `msg.setHeader()` and `.setBody()` things are documented as accepting `Object` instances; however, the comparable bits on the builders are described as `Expression` instances... 

... And then it seems like the test is giving a pass with the expressions due to... the tests evaluating the expressions, or something?

(had this in the buffer, for a good chunk of the morning... now seeing https://github.com/Islandora/Alpaca/pull/82/commits/abe052a99a0d367bca28edbe6d0bd57f1be58162)

adam-vessey replied:
Threw together https://github.com/whikloj/Alpaca/pull/1 ... seems to work, continuing to generate derivatives and whatnot, and nodes and files being indexed in Fedora; however... there's a number of 404s popping, between the Alpaca and Milliner logs... I'm not terribly familiar with Fedora, so I'm not terribly sure what's expected to be there...

adam-vessey commented on islandora-connector-derivative/src/main/java/ca/islandora/alpaca/connector/derivative/DerivativeOptions.java:
```diff
@@ -0,0 +1,190 @@
+  @Override
+  public void setApplicationContext(final ApplicationContext applicationContext) throws BeansException {
+    final String[] names = applicationContext.getBeanNamesForType(CamelContext.class);
+    if (names.length == 1) {
```
PMD's complaining about this:

```
/home/adam/repos/Alpaca/islandora-connector-derivative/src/main/java/ca/islandora/alpaca/connector/derivative/DerivativeOptions.java 	183 	Avoid using Literals in Conditional Statements
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_errorprone.html#avoidliteralsinifcondition

---

Looking further... could be a funny/aggravating scenario, if there were somehow multiple contexts found, but it would report that it found none.

whikloj replied:
My understanding is that in a runnable jar Camel app you only [want a single context](https://camel.apache.org/manual/faq/why-use-multiple-camelcontext.html) but I'll change this to something like
```
private static final int MIN_CAMEL_CONTEXTS = 1;
```
```
if (names.length >= MIN_CAMEL_CONTEXTS) {
```
just in case.

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/FcrepoIndexer.java:
```diff
@@ -18,89 +18,50 @@
-
-
     @Override
     public void configure() {
```
PMD's complaining about this now:

```
/home/adam/repos/Alpaca/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/FcrepoIndexer.java 	59 	Avoid really long methods.
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_design.html#excessivemethodlength

... Not sure if it's the additions here pushing it over whatever threshold, or if it would've reported previously.

whikloj replied:
These PMD rules (I think) have just been running in the background. I do try to check them, but they don't cause a failure.

Anyways, the default rule is max 100 lines and I have the function down to 173 lines by removing some blank lines.

Then I created a processor class to perform some of the reusable things like exchange property setting.

However even by moving these things to a separate class, we only save 20 lines.

I could create a bunch of separate processor classes to move other parts of the routes out of this method, but I feel like it would make following the logic of the route harder to understand.

At this point my vote is to increase the limit or dump the rule.

adam-vessey replied:
Could... possibly just suppress the warning more specifically? https://pmd.github.io/latest/pmd_userdocs_suppressing_warnings.html

... as in, looks like this might do  the trick?
```suggestion
    @SuppressWarnings("PMD.ExcessiveMethodLength")
    public void configure() {
```

... including an inline comment with a bit of rationale might make sense, but yeah, I don't see any great issue going that way, if we want.... I don't see breaking up into a number of different private/protected methods which setup the different bits as being excessively useful, over the one bigger method.

adam-vessey commented on islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java:
```diff
@@ -0,0 +1,155 @@
+            stmt.append("> { ");
+        }
+
+        stmt.append("<");
+        stmt.append(encode(subject));
```
PMD's complaining of a few things here:

```
1 	/home/adam/repos/Alpaca/islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java 	99 	Avoid appending characters as strings in StringBuffer.append.
2 	/home/adam/repos/Alpaca/islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java 	99 	StringBuffer (or StringBuilder).append is called consecutively without reusing the target variable.
3 	/home/adam/repos/Alpaca/islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java 	100 	StringBuffer (or StringBuilder).append is called consecutively without reusing the target variable.
```

Linking to:
* https://pmd.github.io/pmd-6.21.0/pmd_rules_java_performance.html#appendcharacterwithchar
* https://pmd.github.io/pmd-6.21.0/pmd_rules_java_performance.html#consecutiveappendsshouldreuse

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java:
```diff
@@ -0,0 +1,155 @@
+        try {
+            return stream(context.resolvePropertyPlaceholders(property).split(token)).map(String::trim)
+                    .filter(val -> !val.isEmpty()).collect(toList());
+        } catch (final Exception ex) {
```
PMD's complaining here:

```
4 	/home/adam/repos/Alpaca/islandora-indexing-triplestore/src/main/java/ca/islandora/alpaca/indexing/triplestore/processors/ProcessorUtils.java 	150 	Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_design.html#avoidcatchinggenericexception

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java:
```diff
@@ -0,0 +1,121 @@
+    }
+    if (maxConcurrentConsumers > 0) {
+      if (builder.length() > 0) {
+        builder.append("&");
```
PMD complaining here:

```
1 	/home/adam/repos/Alpaca/islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java 	75 	Avoid appending characters as strings in StringBuffer.append.
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_performance.html#appendcharacterwithchar

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java:
```diff
@@ -0,0 +1,121 @@
+   *   The modified http endpoint string.
+   */
+  public String addHttpOptions(final String httpEndpoint, final boolean forceAmpersand) {
+    final StringBuilder builder = new StringBuilder();
```
PMD's complaining here:

```
2 	/home/adam/repos/Alpaca/islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java 	96 	StringBuffer constructor is initialized with size 16, but has at least 45 characters appended.
```

Linking to: https://pmd.github.io/pmd-6.21.0/pmd_rules_java_performance.html#insufficientstringbufferdeclaration

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java:
```diff
@@ -0,0 +1,121 @@
+    // Only append ? or & if there is an endpoint.
+    if (builder.length() > 0) {
+      if (httpEndpoint.contains("?") || forceAmpersand) {
+        builder.append("&");
+      } else {
+        builder.append("?");
```
PMD's the appends:

```
3 	/home/adam/repos/Alpaca/islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java 	101 	Avoid appending characters as strings in StringBuffer.append.
4 	/home/adam/repos/Alpaca/islandora-support/src/main/java/ca/islandora/alpaca/support/config/PropertyConfig.java 	103 	Avoid appending characters as strings in StringBuffer.append.
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_performance.html#appendcharacterwithchar

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey commented on islandora-support/src/main/java/ca/islandora/alpaca/support/event/AS2Object.java:
```diff
@@ -0,0 +1,183 @@
+            throw e;
+        }
+        final var filterUrl = Arrays.stream(url).filter(a -> {
+            if (mimetype != null) {
+                if (a.getMediaType() == null || !a.getMediaType().equalsIgnoreCase(mimetype)) {
```
PMD flagging here:

```
5 	/home/adam/repos/Alpaca/islandora-support/src/main/java/ca/islandora/alpaca/support/event/AS2Object.java 	169 	These nested if statements could be combined
```

Linking to https://pmd.github.io/pmd-6.21.0/pmd_rules_java_design.html#collapsibleifstatements

whikloj replied:
Resolved in https://github.com/Islandora/Alpaca/pull/82/commits/5e5d11e3292a7c6458442aacf47b018476640f8a

adam-vessey reviewed (commented):
Noticed PMD flagging things... should we be introducing/adding in some Github actions that check for these?

DonRichards commented:
@whikloj Sorry, I'm about to show how little I understand Alpaca and this pull request but is it possible to have gradle do the "_build the jar and then copy it into the system (somewhere it can access Fedora, Blazegraph, ActiveMQ, the microservices)._"?

whikloj committed:
Fix PMD rule violations

whikloj committed:
Fix GH actions for different branches

whikloj commented:
Ok there are still 2 PMD rule issues we need to make a decision about, but otherwise they are cleaned up.

@DonRichards once Alpaca is finalized we can deploy a build jar to Maven and Github then people deploying it can just download the pre-built jar instead of having to compile it themselves.

whikloj committed:
Ignore 2 PMD rules in specific cases

whikloj committed:
Correctly reference the exchange

adam-vessey committed:
Move to inject the CamelContext (#1)

* Change up instantiation/initialization slightly.

* PostConstruct is allowed to apply to private methods...

... seems like a false-positive in PMD? Got example exception from:
https://stackoverflow.com/a/48679770

* Changing to camel's 3.7.6 LTS.

adam-vessey commented:
So, did realize something with the concurrent consumer stuff: It wouldn't really allow multiprocessing without the additional [`asyncConsumer` parameter](https://camel.apache.org/components/3.13.x/jms-component.html#_component_option_asyncConsumer)... throwing together another little PR to introduce it:  https://github.com/whikloj/Alpaca/pull/2

whikloj commented:
So I think it depends on what you mean by multiprocessing. For instance in our Islandora Legacy I use camel to kick off tesseract, it will run `concurrentConsumers` equal to the number of cores - 1, but each message is processed completely before the a new one is started. I'm not against the PR for including `asyncConsumer`, I'm just not completely clear how it decides when to grab a new item to start processing. Perhaps this is where the difference between `concurrentConsumers` and maxConcurrentConsumers` comes in?

adam-vessey commented:
Possibly... my interpretation of [`concurrentConsumers`](https://camel.apache.org/components/3.13.x/jms-component.html#_endpoint_query_option_concurrentConsumers) vs [`maxConcurrentConsumers`](https://camel.apache.org/components/3.13.x/jms-component.html#_endpoint_query_option_maxConcurrentConsumers) is somewhat analogous to the `-Xms` and `-Xmx` values to the JVM... initial value when starting, and the max to which things can grow if called for.

Without `asyncConsumer`, it seems like only a single thing for each queue can be processed at a time (per instance of consumers listening, anyway... as in, running multiple instances of the jar could allow for multiple, in a round-about way), but that said... and it's entirely possible that I'm missing something, but given [how it's described](https://camel.apache.org/components/3.13.x/jms-component.html#_concurrent_consuming_with_async_consumer), I'm not sure of the value of being able to specify `concurrentConsumers` _without_ also specifying `asyncConsumer=true` offers much... maybe a alleviating some amount of overhead?

whikloj commented:
I think we are in somewhat agreement, the only difference (I think) is that I now believe that without `asyncConsumers` the `concurrentConsumers` would never rise from its initial value and `maxConcurrentConsumers` might be irrelevant.

I do know that you can set `concurrentConsumers` to a number greater than 1 and have multiple messages processed at the same time. Async is (I think) that once the message has been pulled from the JMS queue and handed off (maybe just to the route or maybe to a secondary route) then a new message is grabbed from the JMS queue. In normal processing it would wait until the first message completed the entire route.

I'll give your PR a build today and I have no issue adding it as a configurable option.

adam-vessey commented:
Hmm... yeah, giving things a further poke, and seeing multiple `convert` instances at a time, from Houdini; however... I wonder... is it because of the streaming? Not really familiar with Camel's lifecycle WRT JMS messages... wondering if it's more, when the `to()` gets a response, that it would acknowledge the message (removing from the queue, allowing another thing to be handled), even though it's not quite done of the message?

whikloj commented:
Yeah I think that is why they don't guarantee the order if you use `asyncConsumers`. If you have a route like
```
from("input")
.to("processing")
.to("output");
```
and queue with 
```
A, B, C
```
with a single consumer but `asyncConsumers=true` then depending on how long each thing takes in `processing` the `output` endpoint could see `A, C, B`, `C, A, B` or `C, B, A`...

adam-vessey committed:
Async consumers for multiprocessing (#2)

* Fix up warning about the unclosed app context.

* Slap together async-consumer stuff.

* Add separator for asyncConsumer parameter.

* Update example as suggested.

adam-vessey commented:
Since... I've now got code in this, I'm not really supposed to merge it, so... yeah... but I would if I could. :P

whikloj commented:
Yeah we'll have to drag some other @Islandora/8-x-committers in to this

elizoller commented:
@whikloj / @adam-vessey  - do you want me to give it another round of smoke tests?

jordandukart commented:
Great job @whikloj and @adam-vessey for working through this!

elizoller commented:
sounds like @alxp already got to it here: https://github.com/Islandora-Devops/islandora-playbook/pull/208#issuecomment-988931232

whikloj commented:
@elizoller if you're able to have a look, please do.

alxp commented on example.properties:
```diff
@@ -0,0 +1,64 @@
+connection.timeout=-1
+socket.timeout=-1
+
+# Solr indexer options
```
Should this be "Fedora indexer options"?

whikloj replied:
DOH

alxp commented on islandora-alpaca-app/src/main/java/ca/islandora/alpaca/driver/AlpacaDriver.java:
```diff
@@ -0,0 +1,84 @@
+    public static void main(final String[] args) {
+        final AlpacaDriver driver = new AlpacaDriver();
+        final CommandLine cmd = new CommandLine(driver);
+        //cmd.setExecutionExceptionHandler(new AppExceptionHandler(driver));
```
Commented-out code should usually be removed.

alxp reviewed (changes requested):
Just one question about the "solr" config comment.

whikloj committed:
Code review

seth-shaw-unlv commented:
@alxp , do you want to merge it? Or are you blocked due to codecov?

whikloj commented:
Thank you all for seeing this through. I'm going to try and release it, see how that goes.
